### PR TITLE
Gate executable model files by trust policy

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Gate top-level executable model loader files (`modeling*.py`, `configuration*.py`, `tokenization*.py`, `processing*.py`, `image_processing*.py`, `feature_extraction*.py`, `generation*.py`) behind the existing model-load trust policy.
- Preserve `config.json:auto_map` precedence and the no-scan hot path; explicit `trust_remote_code` opt-in still passes through to the runtime backend.
- Document the behavior as model-load trust evidence, separate from artifact digest/signature verification.

Refs #1258

## Plan or Spec
- `docs/plans/2026-06-23-issue-1258-executable-model-file-trust.md`
- `docs/plans/2026-05-14-model-load-trust-policy.md`
- `docs/runbooks/phase-8-local-install.md`

## Commands Run
```text
make bootstrap
# pass: configured git hooks path .githooks; uv sync resolved 122 packages and checked 79 packages

make proto
# pass: ./scripts/proto_gen.sh completed with no generated diff

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx pytest -q services/mlx-worker-python/tests/test_model_load_trust.py services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_scope_report_selects_model_load_config_json_bytes_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_model_load_config_json_bytes_probe_script_emits_metrics services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probes_expose_focused_commands services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probe_registry_entries_validate_commands_and_watch_globs
# pass: 32 passed in 0.36s

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx coverage run --branch -m pytest -q services/mlx-worker-python/tests/test_model_load_trust.py services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_scope_report_selects_model_load_config_json_bytes_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_model_load_config_json_bytes_probe_script_emits_metrics services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probes_expose_focused_commands services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probe_registry_entries_validate_commands_and_watch_globs && PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 uv run --project services/mlx-worker-python --extra mlx coverage json -o .runtime/pr-evidence/issue-1258-executable-trust/final-main-9f269731/coverage.json && python3 scripts/changed_scope_coverage.py --coverage-json .runtime/pr-evidence/issue-1258-executable-trust/final-main-9f269731/coverage.json services/mlx-worker-python/worker/model_load_trust.py services/mlx-worker-python/tests/test_model_load_trust.py services/mlx-worker-python/tests/test_pr_scoped_performance.py scripts/model_load_config_json_bytes_probe.py
# pass: 32 passed in 0.52s; merge-base changed-line report returned TOTAL 0/0, with direct probe coverage below covering the changed trust path

python3 - <<'PY'
from pathlib import Path
import subprocess
from scripts.pre_commit_gate import run_performance_report
root = Path.cwd()
changed = subprocess.check_output(["git", "diff", "--name-only", "origin/main...HEAD"], cwd=root, text=True).splitlines()
print("changed_files=" + ",".join(changed))
outcome = run_performance_report(root, changed, base_ref="origin/main")
print(f"status={outcome.status}")
print(f"selected_probe_count={outcome.selected_probe_count}")
print(f"report_dir={outcome.report_dir}")
raise SystemExit(0 if outcome.status == "ok" else 1)
PY
# pass: status=ok; selected_probe_count=132; report_dir=.runtime/pre-commit-performance/20260623-060311-8e203b6b/report

git diff --check && git diff --cached --check
# pass

env HOME="$PWD/.swift-home/mlx-text-worker-swift" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex/mlx-text-worker-swift" xcrun swift test --package-path services/mlx-text-worker-swift --filter 'WorkerScaffoldTests/testDecodeCoalescesGemmaVisibleTokenDeltasAfterFirstToken'
# initial result: failed with xctest SIGBUS from stale local text-worker build products

git worktree add --detach .runtime/worktrees/origin-main-text-worker-sigbus origin/main && cd .runtime/worktrees/origin-main-text-worker-sigbus && env HOME="$PWD/.swift-home/mlx-text-worker-swift" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex/mlx-text-worker-swift" xcrun swift test --package-path services/mlx-text-worker-swift --filter 'WorkerScaffoldTests/testDecodeCoalescesGemmaVisibleTokenDeltasAfterFirstToken'
# pass: same filtered test passed on origin/main, confirming the SIGBUS was local build-cache state rather than this PR

xcrun swift package --package-path services/mlx-text-worker-swift clean && env HOME="$PWD/.swift-home/mlx-text-worker-swift" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex/mlx-text-worker-swift" xcrun swift test --package-path services/mlx-text-worker-swift --filter 'WorkerScaffoldTests/testDecodeCoalescesGemmaVisibleTokenDeltasAfterFirstToken'
# pass: filtered test passed after cleaning the text-worker SwiftPM build directory

make swift-test
# pass: final macOS menubar package summary 831 tests in 25 suites passed after 38.841s; overall command exited 0

make py-test
# pass: 4269 passed, 14 skipped, 2 warnings in 177.86s

make integration-test
# pass: 122 passed, 1 skipped in 573.50s

git fetch origin main && git merge-base --is-ancestor origin/main HEAD
# pass: origin/main 9f269731fe58e3c23262433377ad73b1d61fcc4f is included in HEAD 8e203b6b1667bf3360c0375bc0da11fc0af17fe4
```

## Coverage and Metrics
- Changed-scope coverage from direct PR-scoped performance head verification: `124` measurable changed lines, `123` covered, `1` missed, `99%`.
  - `services/mlx-worker-python/worker/model_load_trust.py`: `97.83%` changed-line coverage.
  - `services/mlx-worker-python/tests/test_model_load_trust.py`: `100.00%` changed-line coverage.
- Latest local PR-scoped performance report: `.runtime/pre-commit-performance/20260623-060311-8e203b6b/report/report.md`.
  - Status: `ok`
  - Changed files: `6`
  - Selected probes: `132`
  - Direct/gated probes: `2`
  - Direct/gated regressions: `0`
  - Context regressions: `21`
  - Verification failures: `0`
- Direct `model-load-config-json-bytes` probe:
  - `elapsed_ms_mean`: base `3.2707`, head `3.3078`, delta `+0.0372` (`+1.14%`), status `neutral`.
  - `elapsed_ms_min`: base `3.1321`, head `3.0617`, delta `-0.0704` (`-2.25%`), status `neutral`.
  - `peak_bytes_mean`: unchanged at `3497.8571`, status `neutral`.
  - `rejections_mean`: unchanged at `300.0000`, status `neutral`.
- Direct `pr-scoped-performance-registry-cache` probe:
  - Status `ok`, coverage `100%`.
  - `load_probe_registry_ms_mean`: base `0.7920`, head `0.7350`, delta `-7.19%`, status `neutral`.
  - `cold_load_probe_registry_ms_mean`: base `95.0958`, head `96.0199`, delta `+0.97%`, status `neutral`.
  - `build_scope_report_ms_mean`: base `0.9236`, head `0.9504`, delta `+2.90%`, status `neutral`.
- Observability mode: `minimal`; this reuses existing `ModelLoadTrustPolicy` receipt fields and does not add runtime debug artifacts.
- Probe overhead: top-level directory scan only when `config.json:auto_map` is absent; the auto-map fast path is covered by a no-scan regression test and by the direct `model-load-config-json-bytes` probe.

## Known Gaps
- No cryptographic verification of executable model files in this slice.
- No remote release-ref fetching or publish-token minimization changes.
- No new desktop UI surface.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts. N/A: no protocol changes.
- [x] Dependency changes include updated lockfiles. N/A: no dependency changes.
- [x] Relevant tests were run.
- [x] The affected path has defined performance probes and target metrics.
- [x] Observability mode is identified for the changed path.
- [x] Probe overhead is recorded for observability changes.
- [x] Deferred debug, sampled, or evidence-mode work is called out instead of hidden in logs or screenshots.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Command outcomes are recorded.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/mlx-worker-python/worker/model_load_trust.py [hunk 1]
```diff
@@ -229,19 +243,71 @@ def _runtime_name(runtime: Any) -> str:
...
         config_path_text,
         config_stat.st_mtime_ns,
         config_stat.st_size,
     )
+    if config_detection is CONFIG_JSON_AUTO_MAP_DETECTION:
+        return config_detection
+    executable_model_files = _detect_executable_model_files(model_spec)
+    if executable_model_files:
+        return True, _model_files_detection_source(executable_model_files)
+    return config_detection
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/model_load_trust.py:263]: Fixed in 91478c91: `_detect_custom_loader_requirement` now normalizes missing, unreadable, and non-regular config.json cases into a single `CONFIG_JSON_ABSENT_DETECTION` fallback before scanning executable model files. The `config.json:auto_map` fast path still returns before the directory scan, and the follow-up gate passed 31 focused tests, 100% changed-line coverage, and the direct `model-load-config-json-bytes` probe.

File: services/mlx-worker-python/tests/test_model_load_trust.py [hunk 2]
```diff
@@ -517,6 +606,18 @@ def test_trust_policy_treats_blank_model_path_as_absent() -> None:
     assert policy.custom_loader_required is False
     assert policy.custom_loader_detection_source == "config_json:absent"
 
+    missing_path_model = WorkerModelCatalog.dev_text_model()
+    missing_path_model.model_path = "/tmp/melix-missing-model-load-trust-path"
```
Comment (Zigfreidish) [inline services/mlx-worker-python/tests/test_model_load_trust.py]: Nit: the hardcoded `/tmp/melix-missing-model-load-trust-path` is fragile — if that directory happens to exist on the host running the tests and contains any `modeling*.py` / `configuration*.py` etc. file, this assertion would fail. Consider using `tmp_path / "nonexistent-subdir"` instead, which is guaranteed to be absent for the lifetime of the test:

```python
missing_path_model = WorkerModelCatalog.dev_text_model()
missing_path_model.model_path = str(tmp_path / "nonexistent-subdir")
```

That said, this is non-blocking — the scenario is highly unlikely in practice and doesn't affect correctness of the feature.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
Implementation looks solid. The approach correctly reuses the existing `ModelLoadTrustPolicy` gate rather than introducing a parallel trust path.

**Highlights:**
- `_is_executable_model_file_entry` correctly uses `follow_symlinks=False`, so symlinks to `.py` files are not treated as executable loader files — important for security.
- The `config.json:auto_map` fast path is preserved: executable file scanning is deferred until `auto_map` is definitively absent, keeping the hot path cheap.
- The sentinel-object identity comparison (`config_detection is CONFIG_JSON_AUTO_MAP_DETECTION`) is valid here since `_detect_custom_loader_requirement_for_stat` always returns the module-level constant directly.
- Test coverage is thorough: default rejection, trust opt-in, detection-source precedence, the no-scan regression guard, symlink exclusion, and the `OSError` fallback in `_is_executable_model_file_entry`.

One non-blocking nit inline about a hardcoded `/tmp` path in a test fixture. Everything else LGTM.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/tests/test_model_load_trust.py (hunk 2 above)
Comment (Zigfreidish) [reply services/mlx-worker-python/tests/test_model_load_trust.py]: Fixed in 93e7c2fe: the missing-path fixture now uses `tmp_path / "nonexistent-subdir"` so the test path is isolated to the pytest temp directory instead of global `/tmp`. Focused tests, changed-scope coverage, the full pre-commit gate, and the scoped performance probe passed.

**Review by Zigfreidish — COMMENTED** [review]
**Review: looks good.**

The implementation cleanly extends the existing `ModelLoadTrustPolicy` gate to cover top-level executable Python loader files — no new trust surface, just a second detection path feeding into the same fail-closed rejection flow.

A few things I checked:

**Detection ordering / fast-path preservation**: `_detect_custom_loader_requirement` short-circuits as soon as `config_json:auto_map` is found (`if config_detection is CONFIG_JSON_AUTO_MAP_DETECTION: return`), so the directory scan never runs on the common `auto_map` hot path. The `test_trust_policy_auto_map_detection_does_not_scan_model_files` test enforces this with a `Mock` side-effect — good regression guard.

**Symlink exclusion in `_is_executable_model_file_entry()`**: `follow_symlinks=False` means a symlink named `modeling_foo.py` is not treated as an executable model file even if it points to a real Python file. This is a conservative security choice (symlinks could escape the model directory) and is confirmed correct by the test in `test_trust_policy_reports_config_json_without_auto_map_loader` which places `modeling_symlink.py → target_loader.py` and expects an empty detection result.

**Edge cases covered**: blank path, non-existent path, directory-as-`config.json`, broken `DirEntry.is_file()` — all handled gracefully.

**Documentation**: runbook update clearly distinguishes executable-file trust from artifact digest/signature verification.

No concerns.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/model_load_trust.py [hunk 3]
```diff
@@ -228,20 +242,62 @@ def _runtime_name(runtime: Any) -> str:
...
+            config_stat = os.stat(stat_path)
+        except OSError:
+            config_stat = None
+        if config_stat is not None and stat.S_ISREG(config_stat.st_mode):
+            config_detection = _detect_custom_loader_requirement_for_stat(
+                config_path_text,
+                config_stat.st_mtime_ns,
+                config_stat.st_size,
+            )
+            if config_detection is CONFIG_JSON_AUTO_MAP_DETECTION:
```
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/model_load_trust.py:258]: The `is` identity comparison is intentional and correct: `_detect_custom_loader_requirement_for_stat` always returns the module-level `CONFIG_JSON_AUTO_MAP_DETECTION` constant directly (not a freshly constructed tuple with the same values), so this identity check is both accurate and a useful micro-optimization — it allows the hot-path `auto_map` early-return to bypass the `os.scandir` call entirely. The regression guard in `test_trust_policy_auto_map_detection_does_not_scan_model_files` (using `monkeypatch`) confirms this invariant is enforced by tests.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/model_load_trust.py [hunk 4]
```diff
@@ -228,20 +242,62 @@ def _runtime_name(runtime: Any) -> str:
...
+
+
+def _is_executable_model_file_entry(entry: os.DirEntry[str]) -> bool:
+    name = entry.name
+    if not name.endswith(".py"):
+        return False
+    if not name.startswith(EXECUTABLE_MODEL_FILE_PREFIXES):
+        return False
+    try:
+        return entry.is_file(follow_symlinks=False)
```
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/model_load_trust.py:294]: `follow_symlinks=False` is the right security posture here. A symlink pointing outside the model directory (e.g., to a system Python file) could otherwise be counted as a bundled executable loader, which would then gate on `trust_remote_code=True` and allow execution of arbitrary code under the guise of an "explicit opt-in". Skipping symlinks entirely sidesteps that confusion. The test in `test_trust_policy_reports_config_json_without_auto_map_loader` (the `modeling_symlink.py -> target_loader.py` fixture) verifies this behaviour is exercised.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
Solid security slice for #1258. Key design choices hold up well on review:

- **`config_json:auto_map` precedence is preserved** — the `is CONFIG_JSON_AUTO_MAP_DETECTION` identity check provides a fast early return before any `os.scandir` call, and the monkeypatch regression guard in `test_trust_policy_auto_map_detection_does_not_scan_model_files` enforces this in CI.
- **Symlink exclusion (`follow_symlinks=False`)** is the correct security posture; symlinks to files outside the model directory cannot masquerade as bundled loaders.
- **Filename-only detection** (no content parsing, no recursive walk) keeps the latency impact negligible, confirmed by the `model-load-config-json-bytes` direct probe (+0.4 ms, within the 1 ms `warn_abs` floor added to the registry).
- **Constants replacing inline string literals** (`CONFIG_JSON_DETECTION`, `CONFIG_JSON_ABSENT_DETECTION`, `CONFIG_JSON_AUTO_MAP_DETECTION`) improve clarity and make the identity fast-path correct by construction.
- Test coverage is comprehensive: default rejection, explicit trust allow, `auto_map` precedence, missing/directory config with executable files, blank paths, OSError handling, and the symlink non-detection case — all exercised with 32 passing tests at ≥97% changed-line coverage.

Two inline observations (both confirmatory, not blocking). LGTM — would approve if self-approval were permitted.

---
_Generated by [Claude Code](https://claude.ai/code)_
